syntax = "proto3";

package guardian_agent;

option go_package = "guardianagent";

message DirFd {
    oneof form {
        int32 fd = 1;
        string path = 2;
    }
}

message Fd {
    int32 fd = 1;
}

message OutBuffer {
    int32 len = 1;
}

message Argument {
    oneof arg {
        int32 int_arg = 1;
        string string_arg = 2;
        DirFd dir_fd_arg = 3;
        Fd fd_arg = 4;
        bytes bytes_arg = 5;
        OutBuffer out_buffer_arg = 6;
    }
}

message Operation {
    int32 syscall_num = 1;
    repeated Argument args = 2;
}

message CredentialRequest {
    Operation op = 1;
    Challenge challenge = 2;
}

message Credential {
    Operation op = 1;
    Challenge challenge = 2;
    bytes signer_nonce = 3;
    string signature_format = 4;
    bytes signature_key = 5;
    bytes signature = 6;
}

message CredentialResponse {
    enum Status {
        APPROVED = 0;
        DENIED = 1;
    }
    Status status = 1;
    Credential credential = 2;
}

message ChallengeRequest {}

message Challenge {
    bytes server_nonce = 1;
    string server_hostname = 2;
    repeated bytes server_public_keys = 3;
}

message ElevationRequest {
    Operation op = 1;
    Credential credential = 2;
}

message ElevationResponse {
    int32 errno_code = 1;
    repeated Argument results = 2;
}
